### PR TITLE
refactor and add event system

### PR DESCRIPTION
**Pull request by mrsoftware**

### CODE REVIEW

File: logger.go [hunk 1]
```diff
@@ -37,46 +37,46 @@ const (
...
-	i.log(InfoLogLevel, message, args...)
-}
+func (l *EventLogger) log(level logLevel, message string, args ...interface{}) {
+	if !l.IsEnable(level) {
+		return
+	}
 
-func (i *IOLogger) Errorf(message string, args ...interface{}) {
-	i.log(ErrorLogLevel, message, args...)
+	fmt.Fprintf(l.writer, "[Vabastegi] ["+level.String()+"] "+message+"\n", args...)
```
Comment (menkaff) [inline logger.go:62]: Do we need to ensure thread safety here by using a mutex?

File: vabastegi.go [hunk 2]
```diff
@@ -8,52 +8,57 @@ import (
...
 
 // Builds the dependency structure of your app.
-func (a *App[T]) Builds(ctx context.Context, providers ...Provider[T]) error {
+func (a *App[T]) Builds(ctx context.Context, providers ...Provider[T]) (err error) {
+	startAt := time.Now()
+
+	a.options.EventHandlers.Publish(&OnBuildsExecuting{BuildAt: startAt})
+
+	defer func() {
+		a.options.EventHandlers.Publish(&OnApplicationShutdownExecuted{Runtime: time.Now().Sub(startAt), Err: err})
```
Comment (menkaff) [inline vabastegi.go]: We can use time.since() instead

File: logger.go (hunk 1 above)
Comment (mrsoftware) [reply logger.go:62]: we don't have any shared resources to change, so why mutex?

Comment (menkaff) [reply logger.go:62]: Yes, you are right. It's useless.

**Review by menkaff — APPROVED** [review]

**Review by menkaff — APPROVED** [review]

**Pull request merged**